src/map/src/combat/combat.cpp

```diff
@@ -215,56 +222,135 @@ void Combat::update(EntitySystem& entitySystem, Entity entity, uint32_t dt) {
+
+          float dmgModifier = (attack_log.value_ + life.maxHp / 15.0f + 30);
+          auto expCalc = ((level.level + 3) * give_exp * dmgModifier);
+
+          if(level_difference <= 3)
+            xp_out = expCalc * server_rate / life.maxHp / 370.0f;
+          else if(level_difference < 9)
+            xp_out = expCalc * server_rate / life.maxHp / (level_difference + 3) / 60.0f;
+          else
+            xp_out = expCalc * server_rate / life.maxHp / (level_difference + 3) / 180.0f;
+
+          if(xp_out < 1)
+            xp_out = 1;
+
+          attackerLevel.xp += xp_out;
+          // End exp out algo
+
+          auto exp_required = get_exp_to_level(attackerLevel.level+1);
+          while (attackerLevel.xp >= exp_required) {
+            int32_t new_exp = attackerLevel.xp - exp_required;
+            attackerLevel.xp = new_exp;
+            ++attackerLevel.level;
+            exp_required = get_exp_to_level(attackerLevel.level+1);
+
+            //TODO:: adjust our hp and mp based on our level here
+            //TODO:: give the player attr and skill points
+            auto p = SrvLevelup::create(attack_log.attacker_, attackerLevel.level, attackerLevel.xp, 0, 0);
+            entitySystem.send_to(attacker, p);
+          }
+
+          int16_t current_stamina = 0;
+          if(entitySystem.has_component<Component::Stamina>(attacker) == true) {
+            auto& attackerStamina = entitySystem.get_component<Component::Stamina>(attacker);
+            current_stamina = attackerStamina.stamina;
+          }
+
+          auto p = SrvSetExp::create(attackerLevel.xp, current_stamina);
+          p.set_source_id(basicInfo.id);
+          entitySystem.send_to(attacker, p);
+
+          xp_out = 1;
+          level_difference = 0;
+        }
+
         entitySystem.add_timer(5s, [entity](EntitySystem& entitySystem) { entitySystem.delete_entity(entity); });
       }
-      
+
       // remove components that we can't have if we are dead!
       entitySystem.remove_component<Component::Combat>(entity);
       entitySystem.remove_component<Component::Target>(entity);
       entitySystem.remove_component<Component::Destination>(entity);
     }
   }
-  
-  // Check to see if we have a target component
+
   if(values.combatDt > 0)
     values.combatDt -= dt;
 
+  // Check to see if we have a target component
   if(entitySystem.has_component<Component::Target>(entity) == true)
   {
     auto& target = entitySystem.get_component<Component::Target>(entity);
     //const auto& targetBasicInfo = entitySystem.get_component<Component::BasicInfo>(target.target);
-    const auto& targetLife = entitySystem.get_component<Component::Life>(target.target);
-    
+
     if(((entitySystem.has_component<Component::Npc>(target.target) == false && entitySystem.has_component<Component::Mob>(target.target) == true) ||
       (false))) // TODO:: Check if this map has PVP turned on and the target player isn't on my team
```

Comment:
maybe we want to have a function to check if the target is valid within a map

Reply:
Will do this in next combat pass